torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -522,17 +522,17 @@ void TensorPipeAgent::join() {
 }
 
 void TensorPipeAgent::shutdownImpl() {
+  // This will close all the pipes and listeners, invoke all callbacks with
+  // errors, turn down the I/O event loops and wait for everything to terminate.
+  context_->join();
```

Comment:
Why this is before `threadPool_.waitWorkComplete();` and `join`? What if there are pending sends in the `threadPool_`? Even in the `graceful=False` case (i.e., no `join`), we might want to deliver all messages in the best effort manner.

Reply:
My reasoning here was that if we did this the other way around, by first waiting on the thread pool, then in some cases we may never shutdown: in particular, if there is a continuous stream of incoming requests, the thread pool would just always be busy and the wait would never return. By first closing the TP context, however, we close all pipes and thus we close the "tap" through which new requests may come in. I do agree however that it makes it much less "graceful". If you think that we should still be somewhat graceful also in this case I'll swap the order.

Reply:
I see. I don't see a reason for using `graceful=False` in OSS with our current implementation. So, it should be fine to land this as-is.